lib/vega_lite/data.ex

```diff
@@ -0,0 +1,231 @@
+  """
+  def heatmap(vl, data, fields) do
+    fields = Enum.map(fields, &heatmap_defaults/1)
+
+    if fields[:text],
+      do: annotated_heatmap(vl, data, fields),
+      else: chart(vl, data, :rect, fields)
+  end
+
+  defp annotated_heatmap(vl, data, fields) do
+    text_fields = [text: fields[:text], x: fields[:x], y: fields[:y]]
+    fields = List.keydelete(fields, :text, 0)
+    used_fields = fields |> Keyword.values() |> used_fields()
+
+    vl
+    |> Vl.data_from_values(data, only: used_fields)
+    |> Vl.layers([layer(data, :rect, fields), layer(data, :text, text_fields)])
+  end
+
+  defp heatmap_defaults({field, [{:field, _col} | _] = opts}) when field in [:x, :y] do
+    {field, Keyword.put_new(opts, :type, :nominal)}
+  end
+
+  defp heatmap_defaults({field, col}) when field in [:x, :y] do
+    {field, [field: col, type: :nominal]}
+  end
+
+  defp heatmap_defaults({field, [{:field, _col} | _] = opts}) when field in [:color, :text] do
+    {field, Keyword.put_new(opts, :type, :quantitative)}
+  end
+
+  defp heatmap_defaults({field, col}) when field in [:color, :text] do
+    {field, [field: col, type: :quantitative]}
+  end
+
+  defp enconde_mark(vl, [{:type, mark} | opts]), do: Vl.mark(vl, mark, opts)
+  defp enconde_mark(vl, mark), do: Vl.mark(vl, mark)
+
+  defp encode_field(schema, cols, field, [{:field, col} | opts]) do
+    opts = Keyword.put_new(opts, :type, type_for(cols, col))
+    Vl.encode_field(schema, field, col, opts)
+  end
+
+  defp encode_field(schema, cols, field, col) do
+    Vl.encode_field(schema, field, col, type: type_for(cols, col))
+  end
+
+  defp layer(data, mark, fields) do
+    cols = columns_for(data)
+    root = Vl.new() |> enconde_mark(mark)
+
+    for {field, col} <- fields, reduce: root do
+      acc -> encode_field(acc, cols, field, col)
+    end
+  end
+
+  defp used_fields(fields) do
+    for field <- fields, field, uniq: true do
+      if is_list(field), do: field[:field], else: field
+    end
+  end
+
+  defp type_for(data, name), do: get_in(data, [Access.filter(&(&1.name == name)), :type]) |> hd()
+
+  defp columns_for(data) do
+    with true <- implements?(Table.Reader, data),
+         data = {_, %{columns: [_ | _] = columns}, _} <- Table.Reader.init(data),
+         true <- Enum.all?(columns, &implements?(String.Chars, &1)) do
+      types = infer_types(data)
+      Enum.zip_with(columns, types, fn column, type -> %{name: to_string(column), type: type} end)
+    else
+      _ -> nil
+    end
+  end
+
+  defp implements?(protocol, value), do: protocol.impl_for(value) != nil
+
+  defp infer_types({:columns, %{columns: _columns}, data}) do
+    Enum.map(data, fn data -> data |> Enum.at(0) |> type_of() end)
+  end
+
+  defp infer_types({:rows, %{columns: columns}, data}) do
+    case Enum.fetch(data, 0) do
+      {:ok, row} -> Enum.map(row, &type_of/1)
+      :error -> Enum.map(columns, fn _ -> nil end)
+    end
+  end
+
+  defp type_of(%mod{}) when mod in [Decimal], do: :quantitative
```

Comment:
I am almost sure we have this logic elsewhere, don't we? Maybe it is on Kino.VegaLite though?